libraries/chain/controller.cpp

```diff
@@ -3030,19 +3030,183 @@ struct controller_impl {
    }
 
    void integrate_received_qc_to_block(const block_id_type& id, const signed_block_ptr& b) {
-#warning validate received QC before integrate it: https://github.com/AntelopeIO/leap/issues/2071
-      // extract QC from block extensions
+      // extract QC from block extension
+      const auto& block_exts = b->validate_and_extract_extensions();
+      if( block_exts.count(quorum_certificate_extension::extension_id()) == 0 ) {
+         return;
+      }
+      const auto& qc_ext = std::get<quorum_certificate_extension>(block_exts. lower_bound(quorum_certificate_extension::extension_id())->second);
+      const auto& new_qc = qc_ext.qc.qc;
+
+      const auto bsp = fork_db_fetch_bsp_by_num( qc_ext.qc.block_height );
```

Comment:
Should search on branch of `id`, not `head->id()`

Reply:
Thanks.